### PR TITLE
dataJoin inserts elements in an order consistent with the bound data

### LINKED ISSUE #528: Order of elements created by data join should match the order of their corresponding datums
The elements in data join's enter selection are appended, so their order may not correspond to the order of the data.

Rather than [appending](https://github.com/mbostock/d3/wiki/Selections#append) the elements, they could be [inserted](https://github.com/mbostock/d3/wiki/Selections#insert) (using `selection.insert()`), to maintain order.

> For enter selections, the before selector may be omitted, in which case entering elements will be inserted immediately before the next following sibling in the update selection, if any. This allows you to insert elements into the DOM in an order consistent with bound data. Note, however, the slower selection.order may still be required if updating elements change order.

Source: https://github.com/mbostock/d3/wiki/Selections#insert

---

This would be particularly useful when the z-index is important, for example the ordering of series in multi.

### PR DESCRIPTION
**Pull request by jleft**
Resolves #528

### CODE REVIEW

File: src/util/dataJoin.js [hunk 1]
```diff
@@ -46,7 +46,7 @@
             // enter
             // when container is a transition, entering elements fade in (from transparent to opaque)
             var enterSelection = updateSelection.enter()
-                .append(element) // <<<--- this is the secret sauce of this whole file
+                .insert(element) // <<<--- this is the secret sauce of this whole file (#528)
```
Comment (jleft) [inline src/util/dataJoin.js]: Is it sufficient referencing 528 here, or would it be better to add a comment with explanation?

Comment (ColinEberhardt) [inline src/util/dataJoin.js]: Personally I'd prefer a comment.

**Comment by ColinEberhardt** [comment]
Thanks for this and thumbs up for unit tests!

I had to read the selection.insert documentation quite closely to see how this change worked. The special behaviour of insert in the case of enter selections is far from obvious!

File: src/util/dataJoin.js (hunk 1 above)
Comment (chrisprice) [inline src/util/dataJoin.js]: I'd explicitly say we're using the special enter-only insert rather than append to help with ordering

Comment (jleft) [inline src/util/dataJoin.js]: Updated with a comment - hopefully the explanation is clear enough?

**Comment by jleft** [comment]
> Thanks for this and thumbs up for unit tests!

No problem, thank you for looking over it :smiley:

**Comment by chrisprice** [comment]
:+1:

**Pull request merged**